programs/mapsoft_convert.1
.TH "mapsoft_convert" "1" "" "slazav@altlinux.org" "mapsoft"
.SH "NAME"
.LP 
Convert geodata to various formats
.SH "SYNTAX"
.LP 
mapsoft_convert <options> <input files> \-o <output file>
.SH "MAIN OPTIONS"
.TP 
\fB\-\-help, \-h\fR
Show help message
.TP 
\fB\-\-out=<file>, \-o <file>\fR
Specify output file (required)
.TP 
\fB\-\-skip=<string>, \-s <string>\fR
Skip some types of data (see below)
.TP 
\fB\-\-rescale_maps <factor>\fR
Rescale map references
.TP 
\fB\-\-shift_maps <x>,<y>\fR
Shift map references
.SH "INPUT FILES"
.TP 
mapsoft XML\-like format
.TP 
old garmin\-utils format (tracks and waypoints)
.TP 
OziExplorer format (tracks, waypoints and map references)
.TP 
geofig format
.TP 
Garmin GPS device via libusb or serial port (use "usb:" or serial device name as a input file name)
.SH "SKIPPING DATA"
.LP 
You may use \fB\-s <string>\fR or \fB\-\-skip=<string>\fR option to skip some data
If string contains letter: 
.TP 
"m" \-\- skip maps 
.TP 
"w" \-\- skip waypoints 
.TP 
"t" \-\- skip tracks 
.TP 
"a" \-\- skip ACTIVE LOG track 
.TP 
"o" \-\- skip all tracks excluding ACTIVE LOG

.SH "INPUT FILES"
.LP
format is determined by file extension:
.TP
\fB *.xml\fR
mapsoft native XML-like format
.TP
\fB *.fig\fR
mapsoft geofig format
.TP
\fB *.wpt, *.plt, *.map\fR
OziExplorer format
.TP
\fB *.zip\fR
zipped OziExplorer files
.TP
\fB *.gpx\fR
GPX format /partial support, tracks and waypoints
.TP
\fB *.gu\fR
old garmin-utils format (tracks and waypoints only)
.TP
\fB gps:\fR
read data from Garmin GPS via autodetected serial device
.TP
\fB <character device>\fR
read data from Garmin GPS via serial device

.SH "OUTPUT FILE"
.LP
Format is determined by file extension:
.TP
\fB *.xml\fR
mapsoft native XML-like format
.TP
\fB *.wpt, *.plt, *.map, *.oe \fR
OziExplorer format (Each track, waypoint set or map reference will be written in a
separate .plt, .wpt or .map file)
.TP
\fB *.zip\fR
zipped OziExplorer files
.TP
\fB *.gpx\fR
GPX format /partial support, tracks and waypoints/
.TP
\fB *.kml\fR
Google KML format /partial support, tracks and waypoints/
.TP
\fB *.kmz\fR
zipped kml
.TP
\fB *.gu \fR
old garmin-utils format (tracks and waypoints only)
.TP
\fB *.tif, .tiff, .jpg, .jpeg\fR
raster image (see below)
.TP
\fB gps:\fR
send data to Garmin GPS via autodetected serial device
.TP
\fB <character device>\fR
send data to Garmin GPS via serial device


.SH "OPTIONS FOR RASTER IMAGE OUTPUT"
.TP 
\fB\-d <name>,\-\-datum=<name>\fR
datum (pulkovo, wgs84, ...)
.TP 
\fB\-p <name>,\-\-proj=<name><fr>
projection (lonlat, tmerc, ...)
.TP 
\fB\-g <AxB+C+D>,\-\-geom=<AxB+C+D>\fR
geometry (in projection units)

.LP 
Default datum is pulkovo, default projection is tmerc..
Default geiometry is calculated from geodata (not map) range.

.TP  
\fB\-\-fig=<file>\fR
write referenced geofig file for image
.TP 
\fB\-\-map=<file>\fR
write OziExplorer map file for image
.TP 
\fB\-\-htm=<file>\fR
write HTML map.
.TP 
\fB\-\-scale <v>\fR
map scale (1e\-5 means 1km per 1cm).
.TP 
\fB\-\-rscale <v>\fR
reversed map scale, 1/scale
.TP 
\fB\-\-dpi <v>\fR
dots per inch
.TP 
\fB\-\-factor <v>\fR
rescale factor (default: 1)

.LP 
By default the scale of resulting map is the same as one of the first.
input map. If there is no maps on input scale=1e\-5,dpi=200.
You may change default value using factor parameter or
rewrite it by dpi and scale (or rscale) parameters
Note: using rscale=100000, dpi=300 and rscale=50000, dpi=150
gave same images but different geofig files.
Note: rscale value overwrites scale.

.TP 
\fB\-\-data_marg=<m>\fR
When detecting geometry of geodata add margings around it (in pixels).

.TP 
\fB\-\-draw_borders=(1|0)\fR
draw input map borders
.TP 
\fB\-\-jpeg_quality=<n>\fR
jpeg quality
.TP 
\fB\-\-max_image=<W,H>\fR
maximum image size (default: 5000,5000)

.TP 
\fB\-\-google=<scale>\fR
add google image of specified scale
.TP 
\fB\-\-ks=<scale>\fR
add ks image of specified scale
.TP 
\fB\-\-google_dir=<dir>\fR
directory for google images
.TP 
\fB\-\-ks_dir=<dir>\fR
directory for ks images
.TP 
\fB\-\-download=(1|0)\fR
turn on downloading of ks and google images
.SH "EXAMPLES"
.TP 
\fBmapsoft_convert /maps/*.map \-o maps.xml\fR
Convert Ozi map files to a single xml
.TP 
\fBmapsoft_convert /dev/ttyUSB0 \-s o \-o data.plt\fR
Download waypoints and active log from GPS device and write them
down in Ozi format files data.plt and data.wpt
.TP 
\fBmapsoft_convert data.plt data.wpt \-o data.kmz\fR
Convert waypoints and track to packed kml format
.TP 
\fBmapsoft_convert maps.xml data.plt \-o mymap.jpg \-\-map=mymap.map\fR
Create jpeg image with map from maps.xml and data from data.plt. Create also a map\-file for this image
.TP 
\fBmapsoft_convert \-\-ks=14 \-\-ks_dir=~/KS \-\-download=1 data.plt \-o mymap.jpg\fR
Create jpeg image with KS image and your dataand map\-file for it. ./KS directory must exist.
.SH "AUTHORS"
.LP 
Vladislav Zavjalov <slazav@altlinux.org>,
Max Ushakov,
Tim Alexeevsky
.SH "SEE ALSO"
.LP 
.TP 
mapsoft (not written yet)
.TP 
mapsoft_vmap (not written yet)
.TP 
mapsoft_mapview (not written yet)


ChangeLog
######
Author: Vladislav Zavjalov <slazav@altlinux.org>
Date:   Sun Oct 12 22:51:58 2008 +0400

    MS3: update DONE

DONE
libzn-utils/scripts/update_from_mp
utils/read_conf.cpp
utils/read_conf.h
utils/read_conf_test.cpp

######
Author: Vladislav Zavjalov <slazav@altlinux.org>
Date:   Sat Oct 11 21:41:25 2008 +0400

    MS3: fix read_conf.cpp

programs/mapsoft_convert.cpp
utils/read_conf.cpp

######
Author: Vladislav Zavjalov <slazav@altlinux.org>
Date:   Sat Oct 11 06:13:25 2008 +0400

    MS3: fix read_conf.cpp

utils/read_conf.cpp

######
Author: Vladislav Zavjalov <slazav@altlinux.org>
Date:   Sat Oct 11 02:13:50 2008 +0400

    MS3: add read_conf.h read_conf.cpp read_conf_test.cpp -- reading options from config files and command line parameters

utils/SConscript
utils/options_read.cpp
utils/read_conf.cpp
utils/read_conf.h
utils/read_conf_test.cpp

######
Author: Vladislav Zavjalov <slazav@altlinux.org>
Date:   Thu Oct 9 12:10:51 2008 +0400

    MS3: add new-style utils/options.h (with operators << >> and test programs)

DONE
utils/SConscript
utils/options.cpp
utils/options.h
utils/options_read.cpp
utils/options_test.cpp

######
Author: Vladislav Zavjalov <slazav@narod.ru>
Date:   Sat Jul 26 20:01:54 2008 +0400

    MS3: remove vector dir (see arc/vector.tgz, arc/vector-znaki.tgz, arc/chfig.tgz)

vector/Makefile
vector/SConscript
vector/add_per_to_fig.cpp
vector/chfig/FORMAT3.2
vector/chfig/README
vector/chfig/XFig.pm
vector/chfig/chfig.pl
vector/chfig/test.fig
vector/chfig/test_out.fig
vector/cnv
vector/cnv_saj
vector/fig_bbx_test.cpp
vector/fig_conv_podm.cpp
vector/fig_maptune.cpp
vector/fig_norm.cpp
vector/mp2fig.cpp
vector/podm/CNV
vector/podm/MAP_PUBL.pl
vector/podm/Makefile
vector/podm/RANGE
vector/podm/bunch.fig
vector/podm/drez.fig
vector/podm/hrab250.fig
vector/podm/isma.fig
vector/podm/jark250.fig
vector/podm/kubr.fig
vector/podm/lutos.fig
vector/podm/olgo250.fig
vector/podm/semisl.fig
vector/podm/sink250.fig
vector/podm/sitava.fig
vector/podm/velja.fig
vector/podm/volg.fig
vector/podm/vypr.fig
vector/test_fig.cpp
vector/test_geofig.cpp
vector/test_mp.cpp
vector/test_poly.cpp
vector/tst_poly.fig
vector/znaki/Makefile
vector/znaki/xfiglib/lines/001-CText.fig
vector/znaki/xfiglib/lines/002-Tunnel.fig
vector/znaki/xfiglib/lines/003-Most-1.fig
vector/znaki/xfiglib/lines/004-Most-2.fig
vector/znaki/xfiglib/lines/005-Most-5.fig
vector/znaki/xfiglib/lines/006-ObrVerh.fig
vector/znaki/xfiglib/lines/007-ObrNiz.fig
vector/znaki/xfiglib/lines/008-ZD.fig
vector/znaki/xfiglib/lines/009-Avto-7.fig
vector/znaki/xfiglib/lines/010-Avto-4.fig
vector/znaki/xfiglib/lines/011-Avto-3.fig
vector/znaki/xfiglib/lines/012-Avto-3-Nepr.fig
vector/znaki/xfiglib/lines/013-AvtoGrunt.fig
vector/znaki/xfiglib/lines/014-ZarGrunt.fig
vector/znaki/xfiglib/lines/015-ProsW.fig
vector/znaki/xfiglib/lines/016-Pros.fig
vector/znaki/xfiglib/lines/017-Tropa.fig
vector/znaki/xfiglib/lines/018-Bound.fig
vector/znaki/xfiglib/lines/019-Dom.fig
vector/znaki/xfiglib/lines/020-LEP-sm.fig
vector/znaki/xfiglib/lines/021-LEP.fig
vector/znaki/xfiglib/lines/022-GAZ.fig
vector/znaki/xfiglib/lines/023-PeresyhR.fig
vector/znaki/xfiglib/lines/024-Ruch-1.fig
vector/znaki/xfiglib/lines/025-Ruch-2.fig
vector/znaki/xfiglib/lines/026-Reka-5.fig
vector/znaki/xfiglib/lines/027-Boloto.fig
vector/znaki/xfiglib/lines/028-Hreb.fig
vector/znaki/xfiglib/lines/029-Ovrag.fig
vector/znaki/xfiglib/lines/030-SuhKan.fig
vector/znaki/xfiglib/lines/031-GorVS.fig
vector/znaki/xfiglib/lines/032-Gor.fig
vector/znaki/xfiglib/lines/033-GorUT.fig
vector/znaki/xfiglib/lines/034-Kontur.fig
vector/znaki/xfiglib/points/001-Triang.fig
vector/znaki/xfiglib/points/002-UrezVody.fig
vector/znaki/xfiglib/points/003-Vysota.fig
vector/znaki/xfiglib/points/004-Text.fig
vector/znaki/xfiglib/points/005-Pam.fig
vector/znaki/xfiglib/points/006-Cerkov.fig
vector/znaki/xfiglib/points/007-Avtobus.fig
vector/znaki/xfiglib/points/008-ZD.fig
vector/znaki/xfiglib/points/009-Kladb.fig
vector/znaki/xfiglib/points/010-Per0.fig
vector/znaki/xfiglib/points/011-Per1.fig
vector/znaki/xfiglib/points/012-Per2.fig
vector/znaki/xfiglib/points/013-Tower.fig
vector/znaki/xfiglib/points/014-Rodn.fig
vector/znaki/xfiglib/points/015-Dom.fig
vector/znaki/xfiglib/points/016-Vdp.fig
vector/znaki/xfiglib/points/017-Por.fig
vector/znaki/xfiglib/points/018-Jama.fig
vector/znaki/xfiglib/points/019-Ohotn.fig
vector/znaki/xfiglib/points/020-Kurgan.fig
vector/znaki/xfiglib/points/021-Skala.fig
vector/znaki/xfiglib/polygon/001-Ostrov.fig
vector/znaki/xfiglib/polygon/002-Voda.fig
vector/znaki/xfiglib/polygon/003-Kladb.fig
vector/znaki/xfiglib/polygon/005-Naselen.fig
vector/znaki/xfiglib/polygon/006-Prom.fig
vector/znaki/xfiglib/polygon/007-OstrLesa.fig
vector/znaki/xfiglib/polygon/008-SvVyr.fig
vector/znaki/xfiglib/polygon/009-StarVyr.fig
vector/znaki/xfiglib/polygon/010-Redkoles.fig
vector/znaki/xfiglib/polygon/011-Led.fig
vector/znaki/xfiglib/polygon/012-Pole.fig
vector/znaki/xfiglib/polygon/013-Les.fig
vector/znaki/zn.fig



programs/mapsoft_convert.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <iomanip>

#include <string>
#include <cstring>
#include <list>

#include <sys/stat.h>
#include <math.h>

#include "geo_io/io.h"
#include "options/m_getopt.h"

using namespace std;

#define OPT1  1  //
#define OPT2  2  // geometry options 
#define OPT3  4  // other raster options
#define OPT4  8  // google/ks options
#define OPT_ALL  (OPT1 | OPT2 | OPT3 | OPT4)

static struct ext_option options[] = {
  {"out",                   1,'o', OPT1, "output file name"},
  {"help",                  0,'h', OPT1, "show this message"},
  {"pod",                   0,  0, OPT1, "show this message as POD template"},
  {"verbose",               0,'v', OPT1, "be verbose\n"},

  {"shift_maps",            1,  0, OPT1, "shift map references, \"x,y\""},
  {"rescale_maps",          1,  0, OPT1, "rescale map references"},
  {"map_nom_brd",           0,  0, OPT1, "set map borders according to map name"},
  {"skip",                  1,'s', OPT1, "skip data, \"wmtao\" (w - waypoints, m - maps, t - tracks, a - active log, o - save tracks)"},
  {"gen_n",                 1,  0, OPT1, "reduce track points to n"},
  {"gen_e",                 1,  0, OPT1, "reduce track points up to accuracy e [meters] (when gen_n and gen_e both used it means: \"remove points while number of points > n OR accuracy < e\""},

  {"geom",          1,  0, OPT2, ""},
  {"datum",         1,  0, OPT2, ""},
  {"proj",          1,  0, OPT2, ""},
  {"lon0",          1,  0, OPT2, ""},
  {"wgs_geom",      1,  0, OPT2, ""},
  {"wgs_brd",       1,  0, OPT2, ""},
  {"nom",           1,  0, OPT2, ""},
  {"google",        1,  0, OPT2, "google tile, \"x,y,z\""},
  {"rscale",        1,  0, OPT2, "reversed scale (10000 for 1:10000 map)"},
  {"dpi",           1,'d', OPT2, "resolution, dots per inch"},
  {"mag",           1,  0, OPT2, "additional magnification"},
  {"swap_y",        0,  0, OPT2, ""},

  {"htm",           1,  0, OPT3, "write html map into file"},
  {"fig",           1,'f', OPT3, "write fig-file"},
  {"map",           1,'m', OPT3, "write map-file"},
  {"draw_borders",  0,  0, OPT3, "draw map borders"},
  {"max_image",     1,  0, OPT3, "don't write images larger then this, \"x,y\", default 1000,1000"},
  {"data_marg",     1,  0, OPT3, "margins around data (works only if no geometry set), pixels"},
  {"jpeg_quality",  1,  0, OPT3, "set jpeg quality"},

  {"ks_zoom",      0,  0, OPT4, ""},
  {"google_zoom",  0,  0, OPT4, ""},
  {"google_dir",   0,  0, OPT4, ""},
  {"ks_dir",       0,  0, OPT4, ""},
  {"download",     0,  0, OPT4, ""},

  {0,0,0,0}
};

void usage(bool pod=false){

  string head = pod? "\n=head1 ":"\n";

  const char * fname = "mapsoft_convert";
  cerr << fname << " -- convert geodata between different formats\n"
       << head << "Usage:\n"
       << "\t"<< fname << " <options> <input files> -o <output file>\n"
       << head << "Input files (format is determined by file extension):\n"
       << "  *.xml -- mapsoft native XML-like format\n"
       << "  *.fig -- mapsoft geofig format\n"
       << "  *.wpt, *.plt, *.map -- OziExplorer format\n"
       << "  *.gpx -- GPX format /partial support, tracks and waypoints/\n"
       << "  *.gu  -- old garmin-utils format (tracks and waypoints only)\n"
       << "  *.zip -- zipped files\n"
       << " gps: --  read data from Garmin GPS via autodetected serial device\n"
       << " <character device> -- read data from Garmin GPS via serial device\n"
       << head << "Output file (format is determined by file extension):\n"
       << "  *.xml -- mapsoft native XML-like format\n"
       << "  *.wpt, *.plt, *.map, *.oe  -- OziExplorer format\n"
       << "      (Each track, waypoint set or map reference will be written in a\n"
       << "       separate .plt, .wpt or .map file)\n"
       << "  *.zip -- zipped OziExplorer files\n"
       << "  *.gpx -- GPX format /partial support, tracks and waypoints/\n"
       << "  *.kml -- Google KML format /partial support, tracks and waypoints/\n"
       << "  *.kmz -- zipped kml\n"
       << "  *.gu  -- old garmin-utils format (tracks and waypoints only)\n"
       << "  *.tif, .tiff, .jpg, .jpeg -- raster image (map, fig, or html wrap can be made)\n"
       << " gps: --  send data to Garmin GPS via autodetected serial device\n"
       << " <character device> -- send data to Garmin GPS via serial device\n"
  ;
  cerr << head << "Options:\n";
  print_options(options, OPT1, cerr, pod);
  cerr << head << "Options for rendering images:\n";
  print_options(options, OPT3, cerr, pod);
  cerr << head << "Options for rendering images, geometry settings:\n";
  print_options(options, OPT2, cerr, pod);
  cerr << head << "Options for rendering images, google/kosmosnimki input:\n";
  print_options(options, OPT4, cerr, pod);

  exit(1);
}

int main(int argc, char *argv[]) {
try{

  if (argc==1) usage();

  vector<string> infiles;
  Options O = parse_options_all(&argc, &argv, options, OPT_ALL, infiles);
  if (O.exists("help")) usage();
  if (O.exists("pod")) usage(true);

  if (!O.exists("out")){
    cerr << "no output files.\n";
    exit(1);
  }

  if (O.exists("verbose")) cerr << "Reading data...\n";

  geo_data world;
  for (vector<string>::const_iterator i = infiles.begin(); i!=infiles.end(); i++)
    io::in(*i, world, O);

  if (O.exists("verbose")){
    cerr << "Map lists: " << world.maps.size()
         << ",  Waypoint lists: " << world.wpts.size()
         << ",  Tracks: " << world.trks.size() << "\n";
  }

  if (O.exists("verbose")) cerr << "Applying filters...\n";
  io::filter(world, O);
  if (O.exists("verbose")) cerr << "Writing data to " << O.get("out", string()) << "\n";
  io::out(O.get("out", string()), world, O);

} catch (const char *err){
  cerr << "ERROR: " << err << "\n";
  exit(1);
}

}

